beamer/chain.py

```diff
@@ -457,19 +406,105 @@ def maybe_challenge(
     log.debug("Withdrew", claim=claim, txn_hash=txn_hash.hex())
+
+
+def handle_request_created(event: RequestCreated, context: Context) -> bool:
+    # Check if the address points to a valid token
+    if context.fill_manager.web3.eth.get_code(event.target_token_address) == HexBytes("0x"):
+        log.info(
+            "Request unfillable, invalid token contract",
+            request_event=event,
+            token_address=event.target_token_address,
+        )
+        return False
+
+    is_valid_request = context.match_checker.is_valid_pair(
+        event.chain_id,
+        event.source_token_address,
+        event.target_chain_id,
+        event.target_token_address,
+    )
+    if not is_valid_request:
+        log.debug("Invalid token pair in request", _event=event)
+        return False
+
+    data = context.request_manager.functions.requests(event.request_id).call()
+    request_data = RequestData.from_chain_data(data)
+
+    req = Request(
+        request_id=event.request_id,
+        source_chain_id=event.chain_id,
+        target_chain_id=request_data.targetChainId,
+        source_token_address=request_data.sourceTokenAddress,
+        target_token_address=request_data.targetTokenAddress,
+        target_address=request_data.targetAddress,
+        amount=request_data.amount,
+        valid_until=request_data.validUntil,
+    )
+    context.requests.add(req)
+    return True
+
+
+def handle_request_filled(event: RequestFilled, context: Context) -> bool:
+    request = context.requests.get(event.request_id)
+    if request is None:
+        return False
+
+    fill_matches_request = (
+        request.id == event.request_id
+        and request.amount == event.amount
+        and request.source_chain_id == event.source_chain_id
+        and request.target_token_address == event.target_token_address
+    )
+    if not fill_matches_request:
+        log.warn("Fill not matching request. Ignoring.", request=request, fill=event)
+        return True
+
+    try:
+        request.fill(filler=event.filler, fill_id=event.fill_id)
+    except TransitionNotAllowed:
+        return False
+    log.info("Request filled", request=request)
+    return True
+
+
+def handle_claim_made(event: ClaimMade, context: Context) -> bool:
+    request = context.requests.get(event.request_id)
+    if request is None:
+        return False
+
+    try:
+        request.claim(event=event, fill_wait_time=context.fill_wait_time)
+    except TransitionNotAllowed:
+        return False
+    log.info("Request claimed", request=request, claim_id=event.claim_id)
+    return True
+
+
+def handle_claim_withdrawn(event: ClaimWithdrawn, context: Context) -> bool:
+    request = context.requests.get(event.request_id)
+    if request is None:
+        return False
+
+    try:
+        request.withdraw()
+    except TransitionNotAllowed:
+        return False
+    log.info("Claim withdrawn", request=request)
+    return True
```

Comment:
I see your point, but my idea was that we wouldn't need an `EventProcessor` for state machine tests at all.

Reply:
There should not be much left to `EventProcessor` after these changes anyway so it should not get in the way too much. In fact, if done properly, it could be much easier for usage in tests, too. Consider that you would only have one entry point into the state machine, as opposed to testing each handler separately (asssuming this is the direction you want to go).

` [events you want to test] -> (state machine) -> [expected output]`

In other words, everything you would want to test with separate handlers, you would still be able to test with the
single-entry-point approach.

Furthermore, you would test both the event processor and event handling at once, whereas if you moved everything out, you would need to have additional tests, which increases the effort. Please also consider maintenance effort here (this was one of the things that killed productivity with Raiden) .

Finally, even ignoring everything above, I think having a minimal public API is quite valuable on its own and should be taken into consideration.

Reply:
> There should not be much left to EventProcessor after these changes anyway so it should not get in the way too much.

But it will require me to run the whole thing in an extra thread. And I won't be able to run just one state transition, e.g. when a state change returns that the state I cannot stop it from being rescheduled.

That *might* not be a problem, but it's not testing the thing we want to test - the pure handling of a certain event.

Reply:
> I do not see why. You would not need to use the entire event processor, or indeed the event processor at all.
What I am trying to say is, you will anyway have an event-dispatch function that is the entry point for the handlers.

That's exactly what you argues for above:
 > There should not be much left to EventProcessor after these changes anyway so it should not get in the way too much. In fact, if done properly, it could be much easier for usage in tests, too.